pgml-apps/pgml-chat/pgml_chat/main.py

```diff
@@ -117,13 +234,112 @@ async def upsert_documents(folder: str) -> int:
     return len(md_files)
 
 
+async def generate_chat_response(
+    user_input,
+    system_prompt,
+    openai_api_key,
+    temperature=0.7,
+    max_tokens=256,
+    top_p=0.9,
+    user_name="",
+):
+    messages = []
+    messages.append({"role": "system", "content": system_prompt})
+
+    chat_history_messages = await chat_collection.get_documents( {
```

Comment:
An implementation more inline with the thesis would combine the first call with this call in a CTE, and follow up with the next call all in 1.